Add DAG copy-empty-like to the C API

alexanderivrii opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

Adds the method ``copy_empty_like`` to the C API.

Addresses #15222.

### Details and comments

I have left some questions on the issue, including whether we should support `VarsMode` and whether we should also expose the `copy_empty_like_with_capacity` variant.

alexanderivrii committed:
initial implementation

alexanderivrii committed:
fixing type

jakelishman commented on releasenotes/notes/qk-dag-copy-empty-like-2b69bf9afd94c1f0.yaml:
```diff
@@ -0,0 +1,12 @@
+---
+features_c:
+  - |
+    Added ``qk_dag_copy_empty_like`` to create a copy of the given DAG,
```
You should be able to do `` :c:func:`qk_dag_copy_empty_like` `` now

alexanderivrii replied:
Done in [85a8872](https://github.com/Qiskit/qiskit/pull/15320/commits/85a88728368514a81165b1b809219c40bb2b5c74).

jakelishman commented on releasenotes/notes/qk-dag-copy-empty-like-2b69bf9afd94c1f0.yaml:
```diff
@@ -0,0 +1,12 @@
+    * duration
+    * all the qubits and clbits, including the registers.
+
+    but without any operations.
```
tbh, we can probably just say "as a mirror of `` :meth:`DAGCircuit.copy_empty_like` ``".

alexanderivrii replied:
Yes, that's much better, thanks, done in [85a8872](https://github.com/Qiskit/qiskit/pull/15320/commits/85a88728368514a81165b1b809219c40bb2b5c74).

jakelishman reviewed (commented):
Thanks for doing this Sasha.  I commented on the issue, mentioning that I think exposing `VarsMode` at this point probably is a sensible idea.

Do you want to do `qk_circuit_copy_empty_like` at the same time, or do that in a separate PR?

jakelishman commented:
Sasha: just bumping this in case you forgot about it.

alexanderivrii committed:
merge with main

alexanderivrii committed:
review comments + QkCircuit.copy_empty_like

alexanderivrii commented:
> Do you want to do qk_circuit_copy_empty_like at the same time, or do that in a separate PR?

Sure, added in [85a8872](https://github.com/Qiskit/qiskit/pull/15320/commits/85a88728368514a81165b1b809219c40bb2b5c74).

Cryoris commented on crates/cext/src/dag.rs:
```diff
@@ -869,3 +869,51 @@ pub unsafe extern "C" fn qk_dag_free(dag: *mut DAGCircuit) {
+/// # Safety
+///
+/// Behavior is undefined if ``dag`` is not either null or a valid pointer to a
+/// ``QkDag``.
```
`dag` is not allowed to be null, is it? I think `const_ptr_as_ref` assumes the pointer is non-null

alexanderivrii replied:
Changed in [0c4b84f](https://github.com/Qiskit/qiskit/pull/15320/commits/0c4b84f6f6d418073cc88937c55d471f5f2b1539).

Cryoris commented on test/c/test_dag.c:
```diff
@@ -434,6 +434,34 @@ static int test_unitary_gates(void) {
+    QkDag *copied_dag = qk_dag_copy_empty_like(dag);
+    size_t num_ops_in_copied_dag = qk_dag_num_op_nodes(copied_dag); // 0
+
+    if (num_ops_in_copied_dag != 0) {
```
For completeness you could check that the original `dag` pointer still has one instruction in it, so we ensure that `qk_dag_copy_empty_like` doesn't affect the original datastructure.

alexanderivrii replied:
Good idea, I have added this extra check both here and in `circuit.rs`, see [0c4b84f](https://github.com/Qiskit/qiskit/pull/15320/commits/0c4b84f6f6d418073cc88937c55d471f5f2b1539).

Cryoris commented on test/c/test_dag.c:
```diff
@@ -434,6 +434,34 @@ static int test_unitary_gates(void) {
+        printf("Expected no operations in the copied-empty-like DAG, but got %zu\n",
+               num_ops_in_copied_dag);
+        result = EqualityError;
+        goto cleanup;
```
We don't need the goto here 🙂

alexanderivrii replied:
This `goto` is now `gone, [0c4b84f](https://github.com/Qiskit/qiskit/pull/15320/commits/0c4b84f6f6d418073cc88937c55d471f5f2b1539).

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -1175,3 +1175,51 @@ pub unsafe extern "C" fn qk_circuit_delay(
+/// * name and other metadata
+/// * global phase
+/// * duration
+/// * all the qubits and clbits, including the registers.
```
`CircuitData` doesn't have a name, metadata or a duration, does it? These are all defined on the Python data structure in addition.

On the `DAGCircuit`, name and metadata here do exist but can only be set via the Python interface at the moment. But with Jake's Python dag -> C dag conversion we could end up copying them here, so there I think the docs are correct

alexanderivrii replied:
Removed name/metadata/duration from description in [10fb938](https://github.com/Qiskit/qiskit/pull/15320/commits/10fb9386979abf9b674334e6e740e22bb43b3e83).

alexanderivrii committed:
merge with main

alexanderivrii committed:
review suggestions

alexanderivrii commented:
I have addressed the review comments and updated the `copy_empty_like` methods to internally use `BlocksMode::Drop`. I did not expose `VarsMode` or the new `BlocksMode` as per Matthew's and Jake's latest comments on the issue, but of course that would be quite simple to do and I would do it if you think this needs to be done in this PR already (maybe to handle `BlocksMode`?)

alexanderivrii committed:
Merge branch 'main' into c_dag_copy_empty_like

alexanderivrii committed:
updating circuit.rs with VarsMode and BlocksMode

alexanderivrii committed:
also updating dag.rs

alexanderivrii commented:
Per discussion, I have updated both `copy_empty_like` methods to handle vars and blocks.

alexanderivrii committed:
Fixing documentation errors

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -1211,3 +1211,98 @@ pub unsafe extern "C" fn qk_circuit_to_dag(circuit: *const CircuitData) -> *mut
+
+/// @ingroup QkCircuit
+///
+/// The mode to use when handling blocks for operations.
```
Could you add that this refers to blocks in control flow instructions, and this is e.g. used when using copy_empty_like? I think we should be a bit more explicit when adding objects to the documentation that don't have very identifiable names 🙂  (Same for `VarsMode` above)

Cryoris commented on crates/cext/src/circuit.rs:
```diff
@@ -1211,3 +1211,98 @@ pub unsafe extern "C" fn qk_circuit_to_dag(circuit: *const CircuitData) -> *mut
+///     uint32_t qubits[1] = {i};
+///     qk_circuit_gate(qc, QkGate_H, qubits, NULL);
+/// }
+/// QkCircuit *copy = qk_circuit_copy_empty_like(qc, QkVarsMode_Alike, QkBlocksMode_Drop);
```
This example could be a little confusing since vars mode and blocks mode don't have any effect. Could you add a comment here explaining that they don't have an effect here, but would have if you had control flow?

Cryoris commented on crates/cext/src/dag.rs:
```diff
@@ -1369,3 +1369,58 @@ pub unsafe extern "C" fn qk_dag_topological_op_nodes(dag: *const DAGCircuit, out
+/// uint32_t qubit[1] = {0};
+/// qk_dag_apply_gate(dag, QkGate_H, qubit, NULL, false);
+///
+/// QkDag *copied_dag = qk_dag_copy_empty_like(dag, QkVarsMode_Alike, QkBlocksMode_Drop);
```
Same here on the comment for vars and blocks mode

alexanderivrii replied:
Thanks @Cryoris, I have added some comments for this and other places in [e3b0c86](https://github.com/Qiskit/qiskit/pull/15320/commits/e3b0c8660182e557c07bc8b09b2dc1cd2b94e631).

alexanderivrii committed:
improved comments based on review

Cryoris commented on crates/cext/src/dag.rs:
```diff
@@ -1394,6 +1394,8 @@ pub unsafe extern "C" fn qk_dag_topological_op_nodes(dag: *const DAGCircuit, out
 /// uint32_t qubit[1] = {0};
 /// qk_dag_apply_gate(dag, QkGate_H, qubit, NULL, false);
 ///
+/// // As the circuit does not contain any control-flow instructions,
```
```suggestion
/// // As the DAG does not contain any control-flow instructions,
```

alexanderivrii replied:
Done in [7d38f80](https://github.com/Qiskit/qiskit/pull/15320/commits/7d38f80b3545f7f9384d9f31fdfc6f5b4c033687): I should be more careful when copy-pasting.

alexanderivrii committed:
circuit -> DAG
